File: src/AnalyticsService.Database/Repositories/DlqRepository.fs
namespace AnalyticsService.Database.Repositories

open AnalyticsService.Database.Model
open AnalyticsService.Database.Model.Domain
open AnalyticsService.Database.Model.Dto
open Dapper.FSharp.PostgreSQL

/// Repository for a dead letter queue.
[<Sealed>]
type DlqRepository() =
    let dlqTable = table'<DeadTopic> "DLQ"

    interface IDlqRepository with
        member this.DeleteDlqItems ids ctx =
            ctx.Conn.DeleteAsync(
                delete {
                    for item in dlqTable do
                        deleteAll
                },
                ctx.Tx
            )

        member this.GetDlqItems conn limit =
            select {
                for _ in dlqTable do
                    selectAll
                    take limit
            } |> conn.SelectAsync<DeadTopicInfo>

        member this.NewDlqTopic topic ctx =
            ctx.Conn.InsertAsync(
                insert {
                    into dlqTable
                    value topic
                },
                ctx.Tx
            )

File: src/AnalyticsService.Database/Repositories/DlqRepository.cs
using System.Data;
using AnalyticsService.Database.Api;
using AnalyticsService.Database.Model.Domain;
using AnalyticsService.Database.Model.Dto;
using AnalyticsService.Database.Queries;
using Dapper;

namespace AnalyticsService.Database.Repositories;

/// <inheritdoc/>
public sealed class DlqRepository : IDlqRepository
{
    /// <inheritdoc/>
    public async Task NewDlqTopic(IDbConnection conn, DeadTopic topic)
    {
        conn.Open();
        using var transaction = conn.BeginTransaction();
        await conn.ExecuteAsync(Query.InsertDlqTopic, topic);
        transaction.Commit();
    }

    /// <inheritdoc/>
    public Task<IEnumerable<DeadTopicInfo>> GetDlqItems(IDbConnection conn, int limit = 1_000)
    {
        conn.Open();
        return conn.QueryAsync<DeadTopicInfo>(Query.GetDlqItems);
    }

    /// <inheritdoc/>
    public async Task DeleteDlqItems(IDbConnection conn, IEnumerable<Guid> ids)
    {
        conn.Open();
        using var transaction = conn.BeginTransaction();
        await conn.ExecuteAsync(
            Query.DeleteDlqItems,
            ids.Select(i => new { Id = i }),
            transaction
        );
        transaction.Commit();
    }
}